src/event_bus/event_bus.py

```diff
@@ -0,0 +1,202 @@
+        """Cancels the subscription associated to a handle
+
+        :param handle: the handle
+        """
+        self.callbacks.pop(handle)
+
+    def _get_callbacks_for_event_type(
+        self, event_type: EventType
+    ) -> Iterator[Callable[[Event], None]]:
+        """Returns an iterator to all callbacks associated with th eenevent type
+
+        :param event_type: The event type
+        :return: The iterator to all the callbacks
+        :yield: a callback
+        """
+        return (
+            callback
+            for callback, e_type in self.callbacks.values()
+            if e_type == event_type
+        )
+
+    def _build_arguments(
+        self, name: str, arg_keys: list[str], args: list, kwargs: dict
+    ) -> dict:
+        """Constructs an argument dict from a list of keys that have to be included
+        a list of argument values and an kwargs-dict
+
+        :param arg_keys: the list of keys
+        :param args: the list of values
+        :param kwargs: the kwargs dict
+        :return: the argument dict
+        """
+        if not len(args) + len(kwargs.keys()) == len(arg_keys):
+            raise TypeError(
+                f"{self.__class__.__name__}.{name}() takes exactly {len(arg_keys)} "
+                f"arguments ({len(args) + len(kwargs.keys())} given)"
+            )
+        arguments = dict(zip(arg_keys, args))
+
+        arg_keys_set = set(key for key in arg_keys if key not in arguments)
+        kwarg_keys_set = set(kwargs.keys())
+        for arg_key in arg_keys_set:
+            if arg_key not in kwarg_keys_set:
+                raise TypeError(
+                    f"{self.__class__.__name__}.{name}() missing required argument '{arg_key}'"
+                )
+        for kwarg_key in kwarg_keys_set:
+            if kwarg_key not in arg_keys_set:
+                raise TypeError(
+                    f"{self.__class__.__name__}.{name}() got an unexpected "
+                    f"keyword argument '{kwarg_key}'"
+                )
+
+        return arguments | kwargs
+
+    def __getattr__(self, name: str) -> object:
+        """This pretends to a caller that the class has the methods
+        defined in EVENT_METHODS. It returns a callable that creates
+        an event and calls all callbacks.
+
+        :param name: the method name
+        :raises AttributeError: when unknown method name is passed
+        :return: a callable
+        """
+        if name not in self.EVENT_METHODS:
+            raise AttributeError(
+                f"'{self.__class__.__name__}' object has no attribute '{name}'"
+            )
+        arg_keys, event_type = self.EVENT_METHODS[name]
+
+        def emit_event(*args: list, **kwargs: dict):
+            arguments = self._build_arguments(name, arg_keys, args, kwargs)
+            event = Event(event_type, arguments)
+            for callback in self._get_callbacks_for_event_type(event_type):
+                callback(event)
+
+        return emit_event
```

Comment:
How exacly does __getattr__ work and when is it called?
Is it called whenever any method or attribute of this class is called to find it and you are overwriting this?
And why does the return value get executed in the other components?

Reply:
`__getattr__` gets called when the caller tries to acces an attribute that doesn't exist. It then creates this attribute and returns it. Because the caller here tries to execute a method, `__getattr__` has to return a callable that then gets executed by the caller.

Reply:
All right. This feels very efficient but also very complicated at the same time. I think it is a good way of doing it, but maybe you could add some comments to prevent some headaches.

Reply:
I didn't want to create a method for every possible event which would effectively all do the same.

But you are right. I will add some comments to clarify the functionality.